### ISSUE TITLE
perf issue with 'ws' in "drop-in" mode

### ISSUE THREAD

**Issue by socmag** [uWebSockets/uWebSockets#199]
Hey, thanks for working on improving ws performance.

I'm testing uWebsocket as a drop-in to replace 'ws.

The test is very basic...

I have a single client that connects and sends 100,000 short MsgPack messages.

Time to completion for ws server: 26,026ms
Time to completion for uWebSockets: 31,060ms

I realize this is a micro-benchmark, and kind of meaningless in isolation, but it is worth noting.

**Comment by SEAPUNK** [comment]
Do you have the benchmark source code for us to test it with?

**Comment by socmag** [comment]
Yup, I'm at lunch now.  Give me a few

**Comment by alexhultman** [comment]
You connect 1 client, send 100k short messages and what is considered completion? Echo?

**Comment by alexhultman** [comment]
Is this over network or just connecting to localhost?

**Comment by socmag** [comment]
localhost

**Comment by socmag** [comment]
No not echo.

The first message prints out the time on the server, then I send 100k
messages Then I instruct the server to print the time again. These are real
timings

**Comment by alexhultman** [comment]
So you have any code? What OS is this on?

**Comment by socmag** [comment]
On my way back from lunch. Will get you the code.

It's on Linux Mint Sarah up to date.

**Comment by alexhultman** [comment]
I can definitely see a difference in the two servers.

I checked, and I get to read 20 bytes all the time (I send 100k 20 byte). ws (the client) sends all the messages as separate packages, this is not what you want to do. This is what TCP_NODELAY is about, queuing up small sends into bigger packages. If you would send 100k messages as one package, or maybe a couple packages then µWS would parse them a lot faster than ws (I wouldn't be surprised about a 50x improvement).

Currently I cork the socket on each read event, so this adds two syscalls per read. I guess this is one difference that shows here. I could make the corks more intelligent by only corking when you actually send something, but its such a small difference in perf. The main idea is to not have the client send a lot of small packages.

**Comment by alexhultman** [comment]
You could make polls that trigger all the time take a break and use a timer of 5ms instead. That way each read would be doing far more. But that's just strange and not really something you want in real situations.

**Comment by alexhultman** [comment]
Something like SO_RCVLOWAT could potentially be used...

**Comment by alexhultman** [comment]
I mean I can get 8ms with SO_RCVLOWAT but it doesn't time out so I cannot use it. But the general idea is to not trigger a read on every 20 bytes. You will never get 20 bytes sent like this if the transmitter uses default TCP settings (aka not using TCP_NOWAIT).

**Comment by alexhultman** [comment]
@socmag Can you test with cork/uncork?

ws._socket.cork();
// do all your sends
ws._socket.uncork();

I can get 2ms with this.

**Comment by alexhultman** [comment]
What's even more messed up is how I get 7 seconds in Node.js, while I get 2 ms in the C++ server...

**Comment by alexhultman** [comment]
@socmag I need your code and I need you to test what I've asked you to test. Both ws and uws are in the *seconds*, not milliseconds in my test.

**Comment by socmag** [comment]
Sorry was distracted here... I think you are on to something here with the
comments above.  I'll try it.

I'm going to send you two separate test suites... the original one with
some of my application logic inside, and another test-suite which is a
purely "synthetic benchmark".

First of all, here is the "synthetic benchmark", and an excel graph that we
made from it.

just run *server.js*, then in another console *client.js*

server.js will print out line by line a CSV

Batch Size,  Time For Batch, Messages Per Second

The client starts by sending 1000 messages, then waits for the server to
process them all, at which point it sends 2000 messages, and so on,
incrementing the batch size by 1000.

You will notice in the graph that on Linux at least, there is a steep cliff
at batch sizes of 75k messages (for both 'ws' and uWebSockets) where all of
a sudden we go from processing ~20k messages/sec to around 4,500.   As we
approach batch sizes of 100k messages it drops off even further.  We are
not sure if this is Node GC, or the networking stack.  We don't think this
is anything to do with you guys.

Okay, here is the test.  I'll send you the one with some of our app code in
a sec.  Need to reboot back to linux

**Comment by SEAPUNK** [comment]
Note that Github issues do not show attachments. You need to upload it somewhere, and then provide a link to it.

**Comment by SEAPUNK** [comment]
A decent host is https://conf.6f.io/ -- it deletes all uploads after 24 hours.

**Comment by SEAPUNK** [comment]
Or if you want a more permanent solution, put it all in a Github Gist.

**Comment by socmag** [comment]
(y)

**Comment by socmag** [comment]
LOL!

Yes that is messed up, but I can believe it :/

So I tried the cork/uncork idea in the client.  At first it didn't make any
difference whatsoever, but then I realized the sends are async, and the
cork/uncork's are not - duh.

Like you said, if I go into a cork/uncork cycles on a timer, things work
much better, but yes, it's an odd way to code it.

I totally it would be better if the client wasn't sending so many tiny
messages flooding the networking stack.  That's fine and I can even do that
in my own protocol layer.  It's not a huge deal - I can batch
hundreds/thousands of messages.  My main concern was why uWS (at least in
my app) was taking longer than ws, and I just wanted you to know.

I'm trying a more intelligent cork/uncork thing right now as well.

**Comment by socmag** [comment]
okay, so this "nagle" algorithm is the best general case we could come up
with without a lot more effort.   It does improve things quite a bit, but
it's not spectacular (about 10-20%) at the cost of some latency.

we're going to try using the C/C++ version of the library and see how we
get on with that.  i'm sure it will be much better.

cheers,
-chris

**Comment by SEAPUNK** [comment]
We still don't have any example code from you. Like I said, whatever files you attached to your email, we cannot see or access them.

**Comment by socmag** [comment]
Ah damn, I'll send it to you in the morning, already left work. Sorry about
that.

The good news is we rewrote it in C++ this evening and now we are handling
300,000 messages at 6% CPU on one core within our broker.

Much better and nice work guys!

We are having beer!

cheers,
-chris

**Comment by alexhultman** [comment]
@socmag You want to spend some time on Gitter tomorrow? I still haven't gotten enough info. It's really strange, I'm getting 7 seconds, ws is getting 5 seconds and the C++ server is getting 2 milliseconds. Maybe this is just JavaScript overhead and not at all about networking or anything other. I still don't get if your timings are in seconds or milliseconds? Do you mean 26 seconds or 26 milliseconds? I get 36 milliseconds in the C++ server without corked sends and I get 2 milliseconds when corking sends.

**Comment by alexhultman** [comment]
@socmag OMFG, I just realized I was using permessage-deflate in Node.js! I disabled it, now it runs about the same as the C++ server (which ofc was running without it). If you cork your sends µWS is about 10x faster just by looking at it.

I think this is really not an issue. If you need good throughput you should either cork your sends or use default TCP settings which will automatically group small sends into larger packets.

**Comment by alexhultman** [comment]
I'm closing this as I now have absolutely no problems significantly outperforming ws, and I don't see any reason to try and tinker with the perf when reading 20 bytes per send (this is not how things are supposed to be used anyways).

If you still have problems getting good perf with µWS I would like to see some code.

**Comment by alexhultman** [comment]
Here are my benchmark numbers if you're interested:

Non-corked send: 1100ms (µWS), 900ms (ws)
Corked send: 90ms (µWS), 900ms (ws)

So yeah, you're right - ws is currently a little faster handling non-corked masses of small messages, but if you apply nagles algorithm or cork ontop of this the perf is dramatically to µWS advantage.

**Comment by alexhultman** [comment]
And with C++:

Non-corked: 1050ms (I need to look into this one!)
Corked: 3ms

**Comment by alexhultman** [comment]
I've checked why ws is faster - they are not getting 20 bytes with every read but rather a chunk of 65kb. I tested removing *everything* in my readable event handler and it still executed in 1050ms. It's not possible to be any faster than this if you are going to respond to every readable event. I guess libuv does some kind of coalescing of receives, will check that out. But yeah, ws is already as fast as it can be at 900ms, while you can potentially get down to 2-3ms in µWS.

**Comment by socmag** [comment]
Hah, you know, I wondered if you were doing that. It did cross my mind.

Well that sounds great.

I'm not going to switch back to Node at this point (for he server at least)
since the C++ version is so fast it's not even funny, plua the memory usage
trundles along at 34MB working set and 6% CPU while the Node clients are
utilizing 1.3GB of RAM and occupying 100% of a full core.

I will be using it for some peripheral services on the other hand, so it is
still impprtant.

I will get you the benchmark when I get to work. Currently waiting here at
home for the postman to pick up a package.

**Comment by socmag** [comment]
Good stuff. Yup closed.

Thanks!

**Comment by alexhultman** [comment]
Yeah I guess you are better off with the C++ server then. Hope it works out good. Btw, I just tested a raw net.Socket in Node.js and it actually has this coalescing built-in, it triggers 65k long chunks so it's not even a matter of ws or not ws, it's more of a difference in the way net.Socket works compared to the raw sockets I use.

**Comment by alexhultman** [comment]
@socmag I cleaned up here a bit. I know why this happens now.

We both have done the very classic flaw of benchmarking *the wrong component*. We've not been benchmarking the *server* at all. Just the client. You cannot benchmark µWS with ws as client - that's wrong in so many ways.

Basically, the bottleneck here is ws - it cannot go any faster than 900 ms no matter if you cork or not. If you replace ws with a high perf client (I just wrote a basic one) you will see µWS outperform ws no matter if you cork or not. We are talking 10-300x better perf in µWS.

* ws client (not corked) -> ws server takes 900 ms (bottleneck: ws)
* ws client (not corked) -> µWS server takes 1050 ms (bottleneck ws, added kernel overhead of excessive syscalls)
* ws client (corked) -> ws server takes 900 ms (bottleneck: ws)
* ws client (corked) -> µWS server takes 3 ms (bottleneck: µWS)
* high perf client (not corked) -> µWS server takes 100 ms (bottleneck: syscalls)
* high perf client (not corked, but using TCP_NOWAIT=false) -> µWS server takes 18 ms (bottleneck: syscalls / µWS)
* high perf client (not corked, but using TCP_NOWAIT=false) -> ws server takes 900 ms (bottleneck: ws)

When µWS is fast enough to process the data faster than the speed it is being produced you get thousands of small reads of size 20 bytes. Since ws is the bottleneck, any disturbance of the resources ws makes use of will slow it down. I guess the added 150 ms comes from µWS reading the buffer ws is writing to all the time, I don't know exactly how this works in the kernel but it makes sense. If you disturb the producer which is the bottleneck you will get worse performance. I don't know if this is correct but that's the closest explanation I can think of.

As an example of how obvious this is [that ws is the bottleneck], I added a blocking sleep of 500 ms in the message handler of µWS. This *speeded up* the receive dramatically. That's not how it works. You don't get better performance by sleeping for 500 ms!

So trying this out with the high perf client, I got something like 10 ms in µWS and obviously ws still performing at 900 ms. Just to really make it glass clear I added a 1 nanosecond sleep between my sends in the client and logged the size of the receives in µWS. They all has the size of 20-50 bytes, while still performing the full receive at about 180 ms. Obviously the bottleneck is ws, which makes this issue invalid from the get go. Always use high perf clients when benchmark, otherwise the benchmark is invalid.

**Comment by socmag** [comment]
Guys...

https://gist.github.com/socmag/a2a195923767bb4196c9559e2fc530ba

Sorry it took so long to get you this, people have had me running around
the last couple of days and it has been quite hectic.

For me, 'ws' is faster than 'uws' in uncorked mode, and 'uws' is faster in
corked mode. (Benchmark timings for ws/uws corked/uncorked in the Gist)

Would be interested to know what you get as results...

cheers,
-chris

**Comment by alexhultman** [comment]
Like I said, ws is **never** faster than uws. You are using an extremely slow **client**, that's the problem. Use a faster client (or cork) and you see that uws is 300x faster than ws. You cannot possibly squeeze out anything faster than 900 ms when using ws, while uws does it in 3 milliseconds.

**Comment by alexhultman** [comment]
This is a fast client, use it and see the 300x (we are talking 30000%) perf diff:

```c++
#include <stdio.h>
#include <stdlib.h>

#include <netdb.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <string.h>

#include <chrono>
#include <thread>
using namespace std;

int main(int argc, char *argv[]) {
   int sockfd, portno, n;
   struct sockaddr_in serv_addr;
   struct hostent *server;

   char buffer[1024];
   portno = 3000;
   sockfd = socket(AF_INET, SOCK_STREAM, 0);
   server = gethostbyname("localhost");
   bzero((char *) &serv_addr, sizeof(serv_addr));
   serv_addr.sin_family = AF_INET;
   bcopy((char *)server->h_addr, (char *)&serv_addr.sin_addr.s_addr, server->h_length);
   serv_addr.sin_port = htons(portno);
   if (connect(sockfd, (struct sockaddr*)&serv_addr, sizeof(serv_addr)) < 0) {
      return -1;
   }

   int flag = 1;
   setsockopt(sockfd, IPPROTO_TCP, TCP_NODELAY, (char *) &flag, sizeof(int));


    const char upgradeHeaderBuffer[] = "GET / HTTP/1.1\r\n"
                                       "Upgrade: websocket\r\n"
                                       "Connection: Upgrade\r\n"
                                       "Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==\r\n"
                                       "Host: server.example.com\r\n"
                                       "Sec-WebSocket-Version: 13\r\n\r\n";

    send(sockfd, upgradeHeaderBuffer, sizeof(upgradeHeaderBuffer) - 1, 0);

    recv(sockfd, buffer, sizeof(buffer) - 1, 0);

    int byteSize = 20, framesPerSend = 1;
    unsigned char *framePackBuffer;
    int framePackBufferLength = 0;

    // Fill with random data
    int allocLength = (byteSize + 14) * framesPerSend;
    framePackBuffer = new unsigned char[allocLength];
    for (int i = 0; i < allocLength; i++) {
        framePackBuffer[i] = rand() % 255;
    }

    // Format message frame(s)
    unsigned char *framePackBufferOffset = framePackBuffer;
    for (int i = 0; i < framesPerSend; i++) {
        framePackBufferOffset[0] = 130;
        if (byteSize < 126) {
            framePackBufferLength += byteSize + 6;
            framePackBufferOffset[1] = 128 | byteSize;
            framePackBufferOffset += byteSize + 6;
        } else if (byteSize <= UINT16_MAX) {
            framePackBufferLength += byteSize + 8;
            framePackBufferOffset[1] = 128 | 126;
            *((uint16_t *) &framePackBufferOffset[2]) = htons(byteSize);
            framePackBufferOffset += byteSize + 8;
        } else {
            framePackBufferLength += byteSize + 14;
            framePackBufferOffset[1] = 128 | 127;
            *((uint64_t *) &framePackBufferOffset[2]) = htobe64(byteSize);
            framePackBufferOffset += byteSize + 14;
        }
    }

    for(int i = 0; i < 100000; i++) {
        send(sockfd, framePackBuffer, framePackBufferLength, 0);
    }

   return 0;
}
```

**Comment by socmag** [comment]
I appreciate that. I'm just sending you the benchmark and the numbers as
requested that I am seeing. Slow client or not.  Did you even look at it?
:-)

Anyway as I said, it isn't even a problem anymore since now I am using your
library in C++.

Which rocks by the way. It far outshines WebSocket++

**Comment by socmag** [comment]
Thanks for the fast client!

**Comment by alexhultman** [comment]
I did take a look at it and I can see your results. The problem is, the results are invalid. You are benchmarking the client. ws never were and never will be faster than uws, that is completely *insane*. You need to realize ws is *the slowest* WebSocket library *of all tested*. It's a complete joke to even consider ws being even remotely close to the speed of uws, especially when stressing the parser like this. The parser is 95% of what makes uws fast. It's a zero-copy SIMD beast built directly ontop of the kernel. ws's parser is built in play dough, hello kitty toy script. What you are observing is kernel contemption and nothing else.

**Comment by alexhultman** [comment]
You need to understand that the `send` syscall is going to take longer time to execute if it needs to trigger a readable event each time, right?

When ws is the server, it is not as fast, so it doesn't have time to read in time, so data sent is being buffered.

* send -> trigger read
* send -> does not trigger read, since server has not read yet, buffer up
* send -> does not trigger read, since server has not read yet, buffer up
* recv
* send -> trigger read

vs.

* send -> trigger read
* recv
* send -> trigger read
* recv
* send -> trigger read

Triggering an event on each send is making *the client* slower and this it seems uws is slower. You are observing an excessive event overhead.

**Comment by socmag** [comment]
Oh I get all right, I've looked at the code for your library and it's quite
beautiful.

One of tge problems is that from the current bits available on NPM, there
is no way to use uws as the client (or did you fix that already).

Personally I don't give a damn, however there will be clients (other than
myself), that will use some bullshit code against your library.

My only purpose for even creating this thread was to try and help. At least
now, when someone else cones along you can point art me and say look, this
dude went through the same exercise. There was a difference which I pointed
out, but now I understand more fully.

Also it was certainly a useful exercise for me having only discovered your
work a couple of days ago.

I'm building a serious piece of infrastructure architecture on top of this
thing, and really do appreciate what you have done.

In fact I almost don't want to tell anyone about it because it's that
goodMy little secret. Haha.

Alright, putting this thing to bed. I'll go grab lunch and then test the
new client.

cheers
-chris

**Comment by socmag** [comment]
new timings are here...

10 clients sending 2,000,000 messages each in parallel
For a total of 20,000,000 messages processed

Total time taken: 2641.09

Messages Per Second: 7,572,638

awesome!

that'll do ;-)

**Comment by alexhultman** [comment]
How does ws perform in this test?

**Comment by socmag** [comment]
That was against my C++ server, I just did even better with 100 clients -
broke the 11 million messages per second barrier.  Maxed out my laptop
though.

I'll try it with the node server in a bit and let you know.  I already know
ws will suck.  The last time I tried it, node was consuming 1.6GB of RAM
and thrashing like a son of a bitch.

By the way, the C++ server just stays steady at ~35MB resident memory.

I gave you a shout out on twitter btw.

**Comment by alexhultman** [comment]
You mind testing ws also? For fun.

**Comment by socmag** [comment]
No of course not, I'll do it.

By the way, I saw you had an issue open regarding wanting help testing on
Windows. I can help with that too.

The only thing with that is that the npm bits didn't build out of the box
with Visual Studio installed, and I didn't have time to look into it.

It looked in the tree like you were also working on a client mode as well.
Is that right?

I have been stuck in meetings a lot, but I'll get to it.

**Comment by alexhultman** [comment]
Yep next version will have client support and a new interface.

**Comment by socmag** [comment]
Okay I ran some tests...

tried varying 1 client, 10 clients 100 clients
tried varying messages sent per client @ 50k, 100k and 2 million

pretty much in all cases, uws showed this kind of profile

932,574
924,898
941,530
921,574
972,384
957,395
951,565
955,155
934,230
960,245
961,908
970,073

pretty much in all cases ws had this kind of profile

246,256
215,517
248,787
213,675
248,737
244,130
241,887
236,406
229,357
240,153

Numbers are messages per second.

So in node basically four times faster out of the box using the fast client
that you gave me.

For my C++ server, the numbers are

2,203,166
2,307,043
6,173,465
11,075,824
12,702,170
10,657,072
12,493,203
10,743,498
12,467,487
12,890,398

Basically the more messages and clients I threw at it, the more it scaled.

Note that for these test, I am not round-tripping the incoming messages
back to the client. They are purely client -> server.

Unfortunately I don't have the setup or time to really dig further into it
right now, but clearly uWS is very good - especially native.

cheers,
-chris
